### PR TITLE
feat: return file hash from uploaded object

### PR DESCRIPTION
**Pull request by juliusmarminge**
syncs https://github.com/pingdotgg/uploadthing-infra/pull/556

image

## Summary by CodeRabbit

- **New Features**
  - Enhanced upload functionality with the addition of a `fileHash` property in the upload completion callback, enabling better tracking of uploaded objects.
  - Updated return values for upload functions to include the new `fileHash` property, providing more comprehensive data post-upload.
  - Introduced a limit on the maximum file count for uploads, set to 4.

- **Documentation**
  - Updated TypeScript documentation link for improved guidance on configuring TypeScript in Next.js applications.

- **Bug Fixes**
  - Improved consistency in test cases by validating the presence of the new `fileHash` property across various upload operations.
  - Removed unnecessary console logging in the middleware function for cleaner operation.

### CODE REVIEW

**Comment by juliusmarminge** [comment]
live tests fails cause infra doesn't yet return this field

File: packages/uploadthing/src/internal/shared-schemas.ts [hunk 1]
```diff
@@ -78,6 +78,7 @@ export class UploadedFileData extends FileUploadDataWithCustomId.extend<Uploaded
   key: S.String,
   url: S.String,
   appUrl: S.String,
+  hash: S.String,
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/shared-schemas.ts:81]: Is hash a good name here? fileHash? contentHash?

Comment (markflorkowski) [reply packages/uploadthing/src/internal/shared-schemas.ts:81]: I think fileHash might be better, more descriptive, good call

File: examples/minimal-appdir/src/server/uploadthing.ts [hunk 2]
```diff
@@ -36,7 +36,7 @@ export const uploadRouter = {
   })
     .middleware(({ req, files }) => {
       // Check some condition based on the incoming requrest
-      console.log("Request", req);
+      req;
       //^?
       // if (!req.headers.get("x-some-header")) {
       //   throw new Error("x-some-header is required");
```
[the remark this answers is not in the document]
Comment (markflorkowski) [reply examples/minimal-appdir/src/server/uploadthing.ts:43]: @coderabbitai This is intentional. Meant as an example of what the user might do.

File: examples/minimal-appdir/src/server/uploadthing.ts [hunk 3]
```diff
@@ -57,6 +58,8 @@ export const uploadRouter = {
       file.customId;
       //   ^?
       console.log("upload completed", file);
+
+      return { foo: "bar" };
```
[the remark this answers is not in the document]
Comment (markflorkowski) [reply examples/minimal-appdir/src/server/uploadthing.ts:62]: @coderabbitai  - This is intentional. The `{ foo: "bar" }` is a placeholder for some returned value from the user's logic. Perhaps a db return or some other computed value.

**Review by markflorkowski — APPROVED** [review]

File: examples/minimal-appdir/src/server/uploadthing.ts [hunk 4]
```diff
@@ -32,11 +32,12 @@
     },
     blob: {
       maxFileSize: "8GB",
+      maxFileCount: 10,
     },
   })
     .middleware(({ req, files }) => {
       // Check some condition based on the incoming requrest
-      console.log("Request", req);
+      req;
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply examples/minimal-appdir/src/server/uploadthing.ts]: i think it's included in the docs getting started as twoslash annotations to show type, but not really needed in the examples

**Pull request merged**